libeos-payg/efi.c

```diff
@@ -474,6 +474,26 @@ eospayg_efi_secureboot_active (void)
   return !!content[0];
 }
 
+/* eospayg_efi_setupmode_active:
+ *
+ * Check if the system was booted with SetupMode
+ *
+ * Returns: %TRUE if booted with SetupMode, %FALSE otherwise
+ */
+gboolean
+eospayg_efi_setupmode_active (void)
+{
+  g_autofree char *tname = full_efi_name (GLOBAL_VARIABLE_GUID, "SetupMode");
+  g_autofree unsigned char *content = NULL;
+  int size;
+
+  content = efivarfs_read (tname, &size);
```

Comment:
If you look at the other accessors for specific variables, such as `eospayg_efi_secureboot_active`, they have a branch for `test_mode` rather than unconditionally reading from the real efivarfs.

Personally I think this is wrong – they should go through the same vfunc as `eospayg_efi_var_read` in my opinion, not least so that code that calls them can be meaningfully tested – but this function should at least be consistent with the others and return a sensible value in tests.

Reply:
I think `eospayg_efi_var_read` is unsuitable here.  Because, it invokes [eospayg_efi_name()](https://github.com/endlessm/eos-payg/blob/b3267a888bd00cce4da11026ecb58949fe73f666/libeos-payg/efi.c#L652C28-L652C44), which prepends the EFI variable name with `EOSPAYG_`.  Then, it finds the wrong one and cannot find the file.

Reply:
You can't use `eospayg_efi_var_read` directly, that's true.

If you call `efivarfs_read` which directly reads from `efivarfs` then you need to check the `test_mode` variable and return some sensible value if it is true.

Or you can use the `efi->read` virtual function, like `eospayg_efi_var_read` does, which either reads from some fake test storage or from the real EFI variables depending on the flags passed to `eospayg_efi_init()` when that function was called.

Reply:
> Because, it invokes [eospayg_efi_name()](https://github.com/endlessm/eos-payg/blob/b3267a888bd00cce4da11026ecb58949fe73f666/libeos-payg/efi.c#L652C28-L652C44), which prepends the EFI variable name with `EOSPAYG_`. Then, it finds the wrong one and cannot find the file.

Notice that there are two delete functions, eospayg_efi_var_delete and eospayg_efi_var_delete_fullname. You could add a `eospayg_efi_var_read_fullname`.